core/src/main/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimit.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+import io.netty.channel.EventLoop;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+final class AsyncConcurrencyLimit implements ConcurrencyLimit {
+    private final int maxParallelism;
+    private final int maxPendingAcquires;
+    private final long acquireTimeoutNanos;
+
+    private final Queue<AcquireTask> pendingAcquireQueue;
+    private final AtomicInteger acquiredPermitCount = new AtomicInteger();
+    private final Runnable timeoutTask;
+    private int pendingAcquireCount;
+    private final SafeCloseable semaphorePermit = this::decrementAndRunTaskQueue;
+
+    AsyncConcurrencyLimit(
+            long acquireTimeoutMillis,
+            int maxParallelism,
+            int maxPendingAcquires) {
+        this.maxParallelism = maxParallelism;
+        this.maxPendingAcquires = maxPendingAcquires;
+        this.pendingAcquireQueue = new ArrayDeque<>(maxPendingAcquires);
+        this.acquireTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(acquireTimeoutMillis);
+        this.timeoutTask = new TimeoutTask();
+    }
+
+    /**
+     * Returns the total number of acquired permits.
+     */
+    public int acquiredPermitCount() {
+        return acquiredPermitCount.get();
+    }
+
+    /**
+     * Returns the total number of available permits.
+     */
+    public int availablePermitCount() {
+        return maxParallelism - acquiredPermitCount();
+    }
+
+    @Override
+    public CompletableFuture<SafeCloseable> acquire(ClientRequestContext requestContext) {
+        return Futures.toCompletableFuture(acquire(requestContext, requestContext.eventLoop().newPromise()));
```

Comment:
Can we just create a new `CompletableFuture` and pass it instead of a Netty `Promise`?
```suggestion
        final CompletableFuture future = new CompletableFuture();
        acquire(requestContext, future);
        return future;
```
This way, we don't need to convert a Netty `Promise` to a `CompletableFuture`.